core/server/src/main/java/alluxio/master/journal/ProtoBufJournalFormatter.java

```diff
@@ -40,11 +47,25 @@ public void serialize(JournalEntry entry, OutputStream outputStream) throws IOEx
   @Override
   public JournalInputStream deserialize(final InputStream inputStream) throws IOException {
     return new JournalInputStream() {
+      private final byte[] mBuffer = new byte[1024];
       private long mLatestSequenceNumber;
 
       @Override
       public JournalEntry getNextEntry() throws IOException {
-        JournalEntry entry = JournalEntry.parseDelimitedFrom(inputStream);
+        int firstByte = inputStream.read();
+        if (firstByte == -1) {
+          return null;
+        }
+        int size = CodedInputStream.readRawVarint32(firstByte, inputStream);
+        byte[] buffer = size <= mBuffer.length ? mBuffer : new byte[size];
+        int bytes = inputStream.read(buffer, 0, size);
+        if (bytes < size) {
```

Comment:
This check assumes that the corrupted entry can only appear at the end of a log, I think this is why the `rotateLog` is called above when I/O error happens.

What about validating an entry in this way: reading the next entry's sequence number, and compare with the current one, if the entry is valid, the next sequence number should be exactly 1 + the current one, otherwise, the entry is corrupted. In this way, there is no need to rotateLog in face of I/O error.

But the rotateLog solution looks good to me since it is a rare corner case.